sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,222 @@
+
+    // Check if this export was the result of a force flush.
+    if (was_force_flush_called == true)
+    {
+      // Since this export was the result of a force flush, signal the
+      // main thread that the worker thread has been notified
+      is_force_flush_.store(false);
+    }
+    else
+    {
+      // If the buffer was empty during the entire `timeout` time interval,
+      // go back to waiting. If this was a spurious wake-up, we export only if
+      // `buffer_` is not empty. This is acceptable because batching is a best
+      // mechanism effort here.
+      if (buffer_.empty() == true)
+      {
+        continue;
+      }
+    }
+
+    auto start = std::chrono::steady_clock::now();
+
+    Export(was_force_flush_called);
+
+    auto end      = std::chrono::steady_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    // Subtract the duration of this export call from the next `timeout`.
+    timeout = schedule_delay_millis_ - duration;
+  }
+}
+
+void BatchLogProcessor::Export(const bool was_force_flush_called)
+{
+  std::vector<std::unique_ptr<Recordable>> records_arr;
+
+  size_t num_records_to_export;
+
+  if (was_force_flush_called == true)
+  {
+    num_records_to_export = buffer_.size();
+  }
+  else
+  {
+    num_records_to_export =
+        buffer_.size() >= max_export_batch_size_ ? max_export_batch_size_ : buffer_.size();
+  }
+
+  buffer_.Consume(
+      num_records_to_export, [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+          std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+          ptr.Swap(swap_ptr);
+          records_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+          return true;
+        });
+      });
+
+  if (exporter_->Export(records_arr) != ExportResult::kSuccess)
+  {
+    // Indicate Error: "[Batch Log Processor]: Failed to export a batch"
+  }
+
+  // Notify the main thread in case this export was the result of a force flush.
+  if (was_force_flush_called == true)
+  {
+    is_force_flush_notified_.store(true);
+    // Notifies the thread
+    force_flush_cv_.notify_one();
+  }
+}
+
+void BatchLogProcessor::DrainQueue()
+{
+  while (buffer_.empty() == false)
+  {
+    Export(false);
+  }
+}
+
+// Note: Timeout functionality is currently not implemented
+bool BatchLogProcessor::Shutdown(std::chrono::microseconds timeout) noexcept
+{
+  // Atomically checking whether value of is_shutdown_ is false
+  // and setting it to true
+  if (is_shutdown_.exchange(true) == false)
+  {
+    // notifies worker thread that shutdown has been called
+    cv_.notify_one();
+    // wait until worker thread completes current export
+    worker_thread_.join();
+    // calls the exporter to shutdown
+    if (exporter_ != nullptr)
+    {
+      return exporter_->Shutdown();
+    }
+    return true;
+  }
+  return false;
```

Comment:
You have a race condition on the "else" here for your destructor.  Scenario:

1. Someone manuals calls `Shutdown` on a thread.
2. The thread owning the BatchLogProcessor call the destructor prior to the `Shutdown` call being complete.
3. Your memory is deleted WHILE you're trying to shutdown.

What I'd recommend is to do the following:

1. Create a "is_shutdown_" that is *only* true IFF shutdown logic is complete.
2. Create a new "latch" "is_shutting_down_" which is true IFF shutdown logic is IN PROGRESS.
3. Create a new condition where you can park `Shutdown()` calls to wait for shutdown.

You'd have logic something like:

```
while (!is_shutdown.load()) {
   if (is_shutting_down.exchange(true)) {
      doTheShutdownThing();
      is_shutdown.store(true);
      waiting_on_shutdown.notify_all();
      is_shutting_down.store(false);
   } else {
     waiting_on_shutdown.wait(timeout);
   }
}
```

Note: I likely didn't get all my atomic arcana correct here, just nudging in the right direction.  I think if you'd rather use a big ole' mutex + wait for the whole shutdown method, that'd be fair for an initial impl and someone can optimise later.   Better to be correct than "Segmentation Fault"

Reply:
Yeah, I agree, with the initial implementation, I think it would be simpler to use the lock, so people who are looking at this later won't have bugs they need to fix. I can definitely look into optimizing this further in a future PR

Reply:
Right, I see that race condition is definitely a problem. With this initial implementation, I agree it would be simpler to use the lock. I've changed that in the latest commit